### PR TITLE
Cst rewrite

### PR DESCRIPTION
**Pull request by srghma**
WIP

https://github.com/purescript-codegen/purescript-ps-ast/issues/1

### CODE REVIEW

**Comment by flip111** [comment]
Hey thanks a lot for your pull request. Could you keep a summary in the first post of what you are changing? Right now it's a lot of changes and i'm not sure what to look for.

**Comment by srghma** [comment]
hi @flip111, I have changed types to the types from purescript-cst 

http://hackage.haskell.org/package/purescript-0.13.6/docs/src/Language.PureScript.CST.Types.html#Declaration

all types from there are added

but not all are printed http://hackage.haskell.org/package/purescript-0.13.6/docs/src/Language.PureScript.CST.Types.html#Declaration

the `Expected.txt` was added to some tests, but these tests are not yet implemented

tests are working https://travis-ci.org/github/srghma/purescript-ps-ast/builds/685609928#L3307

in readme you can find

```
Objectives:
- types are copy of purescript-cst
- parentheses are added automatically, there is no need to use ExprParens or TypeParens (they are added by purescript parser on parsing phase)
```

the only difference from purescript-cst is the `comments` field in `Declaration`s

______

I think everything that I needed is implemented and I'm going to test how it works in https://github.com/srghma/purescript-fernet

**Comment by srghma** [comment]
the functions in `Imports` module are commented now, this module was adding imports from fully qualified names in type and value level during printing (and was making fully qualified names in type and value level non-qualified / was simplifying it)

this is cool feature, maybe in future we will re-add it

**Comment by flip111** [comment]
Looks great so far, give us a notice when you think it's finished and we can do a proper review.

> but not all are printed

Do you mean there is no printer code the stringify some types? I know the purescript compiler doesn't have a full pretty printer. There is this project https://gitlab.com/joneshf/purty which has a pretty printer, but i'm not sure how much it is compatible with the code from the purescript compiler.

> the only difference from purescript-cst is the `comments` field in `Declaration`s

What is the change in the comments field? I think to be able to print comments would be very useful because in case not all typescript definition information (this repo is supposed to work together with purescript-dts later) can be transformed into purescript, it can then be put in some comments at least.

**Comment by paluh** [comment]
@srghma Thanks a lot for your huge contribution. I'm not sure if it is the best place to ask but we have decided that it would be really cool if you join our independent purescript-codegen org team, so I can assure you that you are very welcome :-)
Would you like to join the team?

> tests are working https://travis-ci.org/github/srghma/purescript-ps-ast/builds/685609928#L3307

This is huge!
P.S. For sure we have to cleanup some warnings :-P

> I'm not using purescript-matryoshka (maybe this is bad)

From my perspective it is really good choice. I think that it is better to preserve consistency with the original library than to reach for fancier API in this case.

> I think everything that I need is implemented and I'm going to test how it works in https://github.com/srghma/purescript-fernet

I'm not sure when I find time but I want to cleanup read-dts once again and rewrite codegen for MUI as well using this new API of yours.

**Comment by srghma** [comment]
@flip111

> Do you mean there is no printer code the stringify some types? 

yes

sorry, wrong links, updated [1](https://github.com/srghma/purescript-ps-ast/blob/13f69655db81e4a503442bb9b7e67e5b70fdf8aa/src/Language/PS/AST/Printers.purs#L236), [2](https://github.com/srghma/purescript-ps-ast/blob/13f69655db81e4a503442bb9b7e67e5b70fdf8aa/src/Language/PS/AST/Printers.purs#L367-L368)

> > the only difference from purescript-cst is the `comments` field in `Declaration`s
> 
> What is the change in the comments field? 

I have added comments for [now only to the declarations](https://github.com/srghma/purescript-ps-ast/blob/13f69655db81e4a503442bb9b7e67e5b70fdf8aa/src/Language/PS/AST/Types.purs#L134-L143), but in purescript-cst every element can have comments, before or after the element

e.g. SourceToken contains comment

2020-05-13-15:08:31-screenshot

**Comment by srghma** [comment]
@paluh

> Would you like to join the team?


yes, of course! finally)

> P.S. For sure we have to cleanup some warnings :-P

yes, ok, will do

> > I'm not using purescript-matryoshka (maybe this is bad)
> 
> From my perspective it is really good choice. I think that it is better to preserve consistency with the original library than to reach for fancier API in this case.

:+1:

**Comment by flip111** [comment]
> but in purescript-cst every element can have comments, before or after the element

I don't know if this makes sense in this context. But i once wrote an AST that had a terminal type that had support for the piece of source code + a comment. That way i didn't need to implement the comment for each type separately.

**Comment by srghma** [comment]
@flip111 terminal type? could you make an example?

______

I dont think it makes sense to add comments for every type

yes, it could be useful to be able to write someday

```purs
foo = bar {- COMMENT1 -} {- COMMENT2 -} foo {- COMMENT3 -}
```

but I dont need this for now

for now I need comments only at `Declaration`

**Comment by srghma** [comment]
@paluh @flip111 @jvliwanag I think it's ready, could you please make a review?

should we make another repo called `purescript-ps-cst` OR merge to this repo?

**Comment by flip111** [comment]
> @flip111 terminal type? could you make an example?

In the grammer you have non-terminal rules, and terminal rules (tokens). Your tokens can be a String or so. But you can also make it a record of String for the token and a Maybe String for a comment.

> should we make another repo called `purescript-ps-cst` OR merge to this repo?

We can merge here and then rename in to purescript-ps-cst IMO

**Comment by flip111** [comment]
@srghma Do you think we can do the following:

Generate random AST with quickcheck
Print it into source code
Read source code with parser
Verify that AST from parser matches the AST we put in

**Comment by srghma** [comment]
> In the grammer you have non-terminal rules, and terminal rules (tokens). Your tokens can be a String or so. But you can also make it a record of String for the token and a Maybe String for a comment.

that's basically Haskell types, with `Wrapped` and everything else

> Generate random AST with quickcheck
> Print it into source code
> Read source code with parser
> Verify that AST from parser matches the AST we put in

we don't have parser implemented in purescript

**Comment by paluh** [comment]
> (...) I think it's ready, could you please make a review?

@srghma this is great! Please give me "a moment" - I hope to do the review on the next few days... I'm just not sure if I'm able to add anything valuable to this PR ;-)

File: README.md [hunk 1]
```diff
@@ -2,3 +2,10 @@
 
 Currently this package provides an opinionated AST for PureScript which was ripped from purescript-react-basic-mui codegen tool. We hope to turn it into somethig more generic and complete.
 
+TODO:
```
Comment (paluh) [inline README.md:3]: Would you be so kind and drop also the above section (I mean "Currently this package...") which would be out of date with this merge?

File: packages.dhall [hunk 2]
```diff
@@ -121,20 +121,29 @@ let additions =
 let upstream =
       https://github.com/purescript/package-sets/releases/download/psc-0.13.6-20200404/packages.dhall sha256:f239f2e215d0cbd5c203307701748581938f74c4c78f4aeffa32c11c131ef7b6
 
-let overrides = {=}
```
Comment (paluh) [inline packages.dhall:124]: Could you please make a PR from this: https://github.com/srghma/purescript-either/commit/b2c44fcb93b7cff61bf80799da3ed64c5428b6d8
so we can depend on `either` upstream soon?

File: packages.dhall [hunk 3]
```diff
@@ -121,20 +121,29 @@ let additions =
...
-        [ "fixed-points"
-        , "free"
-        , "prelude"
-        , "profunctor"
-        , "transformers"
-        ]
-    , repo = "https://github.com/slamdata/purescript-matryoshka.git"
-    , version = "v0.4.0"
-    }
-  }
```
Comment (paluh) [inline packages.dhall:138]: It seems that we are able to drop `matryoshka` all together. Am I right?

File: src/Language/PS/AST.purs [hunk 4]
```diff
@@ -1,6 +1,5 @@
 module Language.PS.AST
```
Comment (paluh) [inline src/Language/PS/AST.purs]: Should we consider migration to the original namespace: `Language.PureScript.CST`? What do you think Guys?

File: src/Language/PS/AST/Imports.purs_ [hunk 5]
```diff
@@ -39,6 +39,7 @@ declarationImports (DeclValue { value, signature: Just s }) =
   cata typeImportsAlgebra s <> cata exprImportsAlgebra value.expr
 declarationImports (DeclValue { value, signature: Nothing }) =
   cata exprImportsAlgebra value.expr
```
Comment (paluh) [inline src/Language/PS/AST/Imports.purs_]: We should probably drop this file all together as it resides in git history for reference ;-)

File: packages.dhall (hunk 2 above)
Comment (srghma) [reply packages.dhall:124]: there was a pr already https://github.com/purescript/purescript-either/pull/51

File: packages.dhall (hunk 3 above)
Comment (srghma) [reply packages.dhall:138]: yes

File: src/Language/PS/AST.purs (hunk 4 above)
Comment (srghma) [reply src/Language/PS/AST.purs]: done Guys)

**Comment by srghma** [comment]
@paluh thanks for kind words, did as for myself

review notes are resolved)

File: README.md (hunk 1 above)
Comment (paluh) [reply README.md:3]: It is a detail - I'm going to drop it later myself ;-)

**Comment by srghma** [comment]
created https://github.com/purescript-codegen/purescript-ps-cst

**Pull request closed without merging**